[Doc] Refactor ODOP (#6013)
taichi-dev/taichi: docs/lang/articles/advanced/odop1.md

sidebar_position: 2
---

# Objective Data-oriented Programming I

Taichi is a [data-oriented](https://en.wikipedia.org/wiki/Data-oriented_design) programming (DOP) language. However, simple DOP makes modularization hard. To allow modularized code, Taichi borrows some concepts from object-oriented programming (OOP). For convenience, let's call the hybrid scheme **objective data-oriented programming** (ODOP).

The ODOP scheme allows you to organize data and methods into a class and call the methods to manipulate the data in the Taichi scope. Taichi offers two different types of classes that serve this purpose, and they are distinguished by the two decorators `@ti.data_oriented` and `@ti.dataclass`, respectively:

> 1. `@ti.data_oriented`: This decorator is recommended when your data is actively updated in the Python scope (such as current time and user input events) and tracked in Taichi kernels. This class can have native Python objects as members and must be instantiated in the Python scope. This article will discuss this type of class in full detail.

Reviewer: ```suggestion
1. `@ti.data_oriented`: It should be used when your data is actively updated in the Python scope (such as current time and user input events) and tracked in Taichi kernels. This type of class can have native Python objects as members and must be instantiated in the Python scope. This article will discuss this type of class in full detail.
```

> 2. `@ti.dataclass`: A class decorated by `@ti.dataclass` is a wrapper over `ti.types.struct` but offers more flexibility than `ti.types.struct`: You can define Taichi functions as its methods and invoke these methods in the Taichi scope. We will discuss this type of class in the next article.

Reviewer: ```suggestion
2. `@ti.dataclass`: It is a wrapper over `ti.types.struct` but offers more flexibility: You can define Taichi functions as its methods and invoke these methods in the Taichi scope. We will discuss this type of class in the next article.
```
